### PR TITLE
keep compare tab in sync with selection

### LINKED ISSUE #4985: selected commit not kept in sync with compare tab state
I think we're close to cleaning up the old `History` component which might uncover the root cause of this, but this is a regression I uncovered while testing #4874 so ignore that part of the screenshots.

Repro steps:

 - under new History tab, select a commit (whether you're comparing or not doesn't matter)
 - switch from History -> Changes -> History

**Expected:**

 - commit in diff panel appears selected in list

**Actual:**
 
 - no selection persisted, despite diff being active

This is even more noticeable when you're in the middle of a compare, because it's ignores the commit I had selected and reverts to the commit I had chosen before:

### LINKED ISSUE #5069: Selected commit in Compare branch view gets replaced with latest when window regains focus
## Description

When you select a commit in the Compare to Branch view, it's diff is presented in the right part of window. However when the window gets out of focus and then is focused again, this diff is replaced with latest commit's diff. Previously selected commit still looks selected and clicking on it doesn't refresh the diff. One has to select different commit to change presented diff. The issue is only present when a commit from "Behind" list is selected.

## Version

* GitHub Desktop: 1.2.6

* Operating system: macOS 10.13.4

## Steps to Reproduce

1. Go to "Compare to Branch" view
2. Select some branch
3. Select some commit from "Behind" list
4. Get the window out of focus / open different application
5. Refocus the window

### Expected Behavior

The diff should remain same

### Actual Behavior

The diff from the first commit on the branch is presented

### PR DESCRIPTION
**Pull request by shiftkey**
This PR removes the last vestiges of `IHistoryState` by moving selection into a root `ICommitSelection` state and ensuring the Compare tab uses it when displaying it's list of commits. This helps us close out some selection-related bugs that have been lurking for a while.

Fixes #4985
Fixes #5069

- [x] confirm #4985 is fixed
- [x] confirm #5069 is fixed
- [x] 📝 all the public members (old and new)
- [x] :fire: any last History-related code
- test various scenarios
   - [x] when viewing history - scrolling populates with older commits, history remains preserved
   - [x] when viewing history - scroll, select commit, change tabs, return -> commit remains selected
   - [x] when viewing history - scroll, select commit, switch to other app, return -> commit remains selected
   - [x] when viewing compare - scroll, select commit, change tabs, return -> commit remains selected
   - [x] when viewing history - select commit, then compare branch -> first ahead commit selected
   - [x] when viewing history - select commit, then compare branch which is not ahead -> no commit selected
   - [x] when viewing compare - select commit, then switch back to history -> first commit selected

### CODE REVIEW

**Comment by shiftkey** [comment]
Tentatively pulling this into 1.3.0 but will move to a later version if we run out of time

File: app/src/lib/app-state.ts [hunk 1]
```diff
@@ -342,7 +342,7 @@ export enum RepositorySectionTab {
 }
 
 export interface IRepositoryState {
-  readonly historyState: IHistoryState
+  readonly selection: ICommitSelection
```
Comment (iAmWillShepherd) [inline app/src/lib/app-state.ts]: Can this be renamed to `commitSelection`?

File: app/src/lib/app-state.ts [hunk 2]
```diff
@@ -561,19 +561,22 @@ export interface IBranchesState {
...
-  readonly file: CommittedFileChange | null
-}
-
-export interface IHistoryState {
-  readonly selection: IHistorySelection
-
-  /** The ordered SHAs. */
-  readonly history: ReadonlyArray<string>
-
+  /**
```
Comment (iAmWillShepherd) [inline app/src/lib/app-state.ts]: Newline before the start of this comment

Comment (iAmWillShepherd) [reply app/src/lib/app-state.ts]: There are two more instances of this.

**Comment by shiftkey** [comment]
:apple:

**Pull request merged**